src/pip/_internal/req/constructors.py

```diff
@@ -353,6 +354,12 @@ def with_source(text):
     if req_as_string is not None:
         try:
             req = Requirement(req_as_string)
+            # Check that specifiers are PEP-440 compliant
+            for spec in req.specifier:
+                if not isinstance(spec, Specifier):
+                    msg = "Invalid specifier format '{}'".format(str(spec))
+                    deprecated(msg, replacement=None, gone_in=None)
```

Comment:
We should probably move this entire loop into an else block associated with this `try-except`.